File: 백준/Gold/1799. 비숍/비숍.cc
#include <iostream>
#include <vector>
using namespace std;

enum color {BLACK, WHITE};
struct xy {int x, y;};

int n, ans[2];
vector<xy> B[2];
bool diag[2][2][20]; // [color][0]: +, [color][1]: -

void backtracking(color c, int cur, int cnt) {
    if(cur == B[c].size()) {
        if(cnt > ans[c]) ans[c] = cnt;
        return;
    }
    if(cnt + B[c].size() - cur < ans[c]) return;

    if(diag[c][0][B[c][cur].x + B[c][cur].y] or diag[c][1][B[c][cur].x - B[c][cur].y + n]) {
        backtracking(c, cur+1, cnt);
        return;
    }

    diag[c][0][B[c][cur].x + B[c][cur].y] = diag[c][1][B[c][cur].x - B[c][cur].y + n] = true;
    backtracking(c, cur+1, cnt+1);
    diag[c][0][B[c][cur].x + B[c][cur].y] = diag[c][1][B[c][cur].x - B[c][cur].y + n] = false;
    backtracking(c, cur+1, cnt);
}

int main() {
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> n;
    bool isB;
    for(int x=0; x<n; x++) {
        for(int y=0; y<n; y++) {
            cin >> isB;
            if(isB) B[(x ^ y) & 1].push_back({x, y});
        }
    }
    
    backtracking(BLACK, 0, 0);
    backtracking(WHITE, 0, 0);
    cout << ans[BLACK] + ans[WHITE];
}

File: 백준/Gold/1799. 비숍/비숍.py
import sys
readline = sys.stdin.readline

N = int(readline())
board = [list(map(int, readline().split())) for _ in range(N)]
diagonals = [[[False] * (2 * N) for _ in range(2)] for _ in range(2)]

ans = [0, 0]
candidates = [[], []]
for i in range(N):
    for j in range(N):
        if board[i][j] == 1:
            candidates[(i ^ j) & 1].append((i, j))

def backtracking(color, cur, cnt):
    if cur == len(candidates[color]):
        ans[color] = max(ans[color], cnt)
        return
    
    if cnt + len(candidates[color]) - cur <= ans[color]:
        return
    
    x, y = candidates[color][cur]
    if diagonals[color][0][x + y] or diagonals[color][1][x - y + N]:
        backtracking(color, cur + 1, cnt)
        return
    
    diagonals[color][0][x + y] = diagonals[color][1][x - y + N] = True
    backtracking(color, cur + 1, cnt + 1)
    diagonals[color][0][x + y] = diagonals[color][1][x - y + N] = False
    backtracking(color, cur + 1, cnt)

backtracking(0, 0, 0)
backtracking(1, 0, 0)
print(ans[0] + ans[1])